### PR TITLE
New serialization

### PR DESCRIPTION
**Pull request by vweevers**
Do not merge yet.

I can break this PR up but first I wanted to get feedback on the changes as a whole. You can follow the commits, or start with the readme and changelog.

### CODE REVIEW

File: memdown.js [hunk 1]
```diff
@@ -200,6 +205,7 @@ MemDOWN.prototype._iterator = function (options) {
   return new MemIterator(this, options)
 }
 
+MemIterator.prototype._isBuffer =
```
Comment (ralphtheninja) [inline memdown.js]: Why is this even a prototype method? Wouldn't a simple function be more performant? Are there any implementations overriding it?

File: README.md [hunk 2]
```diff
@@ -12,37 +12,104 @@
...
+
+db.put('key', obj, (err) => {
+  obj.thing = 'modified'
+
+  db.get('key', { asBuffer: false }, (err, value) => {
+    console.log(value === obj) // false
+    console.log(obj.thing) // 'original'
+  })
+})
+```
```
Comment (ralphtheninja) [inline README.md]: :+1:

File: memdown.js (hunk 1 above)
Comment (vweevers) [reply memdown.js]: Historically it was there for the use of `bops`. I also vaguely recall some issue with browserify, but I can't find it. I'll try calling `Buffer.isBuffer` directly; if the browser tests pass we should be good.

Comment (vweevers) [reply memdown.js]: => Level/abstract-leveldown#149

**Comment by vweevers** [comment]
@ralphtheninja @juliangruber mostly I want to make sure that you're OK with this new direction (of memdown supporting all JS types). Any thoughts or concerns?

File: README.md [hunk 3]
```diff
@@ -43,6 +33,83 @@ Browser support
...
+    keyAsBuffer: false,
+    valueAsBuffer: false
+  }).on('data', (entry) => {
+    console.log(typeof entry.key) // 'number'
+    console.log(typeof entry.value) // 'boolean'
+  })
+})
+```
+
+Practically speaking, few types are suitable for keys. You can use Buffers or strings, which sort lexicographically, or numbers or Dates, which sort naturally. That's about it.
```
Comment (juliangruber) [inline README.md]: this should be up to the consumer to decide, can we reword this as a hint?

File: README.md (hunk 2 above)
Comment (juliangruber) [reply README.md]: yeah very nice documentation!

File: memdown.js [hunk 4]
```diff
@@ -127,9 +132,15 @@ MemDOWN.prototype._open = function (options, callback) {
...
 }
 
-MemDOWN.prototype._put = function (key, value, options, callback) {
-  if (typeof value === 'undefined' || value === null) value = ''
+MemDOWN.prototype._serializeKey = function (key) {
+  return key
+}
+
+MemDOWN.prototype._serializeValue = function (value) {
+  return value == null ? '' : value
```
Comment (juliangruber) [inline memdown.js]: why would we not store `null`?

Comment (vweevers) [reply memdown.js]: This null behavior is the same as the default `_serializeValue` implementation of `abstract-leveldown`. I wanted to keep this as-is for now, and attack the null/undefined question later (across the board).

Main reason not to store null is because `null` signifies end in Node.js streams. More at https://github.com/Level/abstract-leveldown/issues/120.

File: README.md (hunk 3 above)
Comment (vweevers) [reply README.md]: I'm all for letting the consumer decide, but what other types are there? That sort in a sensible way without needing a custom comparator (which btw, might be a nice feature to have?).

Comment (juliangruber) [reply README.md]: maybe they don't need it to sort at all

File: memdown.js (hunk 4 above)
Comment (juliangruber) [reply memdown.js]: true, I remember now :)

File: README.md (hunk 3 above)
Comment (vweevers) [reply README.md]: If it doesn't sort (specifically: if two keys are equal) then `get()` doesn't work, because `functional-red-black-tree` won't know which node to return.

Comment (vweevers) [reply README.md]: Actually no, I'm wrong there, as `put` will update the node.

Comment (vweevers) [reply README.md]: Suggestions for new wording?

Comment (juliangruber) [reply README.md]: I think we don't need to mention this at all, people won't think about custom key types unless they have something special in mind anyway, in which case they probably know enough about the platform to know the limitations.

Comment (vweevers) [reply README.md]: That's true. On the other hand, if we inform people (newcomers) of the possibilities, we might inspire them :) So I prefer your earlier suggestion of rewording as a hint.

How about this? Move it to the paragraph above it, in a more "optimistic" tone, and also segway into a note about `null` and `undefined` keys (which I now realize is missing):

---

Unlike [`leveldown`], `memdown` does not stringify keys or values. This means that in addition to Buffers, you can store any JS type without the need for [`encoding-down`]. For keys for example, you could use Buffers or strings, which sort lexicographically, or numbers, even Dates, which sort naturally. The only exceptions are `null` and `undefined`. Keys of that type are rejected; values of that type are converted to empty strings.

[`encoding-down`]: https://github.com/Level/encoding-down
[`leveldown`]: https://github.com/Level/leveldown

Comment (juliangruber) [reply README.md]: Yes, this ^

**Comment by ralphtheninja** [comment]
@vweevers Can we get this merged?

**Comment by vweevers** [comment]
Sure. Lemme:

- [ ] Address https://github.com/Level/memdown/pull/106#discussion_r154593421
- [ ] Remove the object keys test
- [ ] Pin `abstract-leveldown` to latest commit
- [ ] Check if test suite needs updating
- [ ] Open an issue to switch to versioned `abstract-leveldown` once published
- [ ] Open an issue to remove `_isBuffer`
- [ ] Squash (not all)

**Comment by vweevers** [comment]
@ralphtheninja do I need to change anything else in the test suite?

**Comment by vweevers** [comment]
The [build failure](https://travis-ci.org/Level/memdown/jobs/325778664) is probably a random connection error:

```
W: Failed to fetch http://www.apache.org/dist/cassandra/debian/dists/39x/InRelease  
Could not connect to www.apache.org:80 (140.211.11.105), connection timed out
W: Failed to fetch http://ppa.launchpad.net/chris-lea/redis-server/ubuntu/dists/trusty/InRelease  
Could not connect to ppa.launchpad.net:80 (91.189.95.83), connection timed out
```

Restarting.

File: memdown.js (hunk 1 above)
Comment (vweevers) [reply memdown.js]: For reference, there is indeed a browserify issue: https://github.com/Level/abstract-leveldown/issues/172
